### DISCUSSION TITLE
Questions regarding FrankenPHP and async processes

### DISCUSSION THREAD

**Discussion by gabrielius914** [php/frankenphp#577] [Q&A] [answered]
Hey, 
1) does the message consumer load the PHP application (services, kernel, etc) each time the message is received (like with the request)?
2) is it possible to improve Symfony Messenger's consumers performance with FrankenPHP worker mode or it only applies for HTTP requests?

**Comment by dunglas** [comment, marked as answer]
Hey,

1. No, it loads it only one time per process.
2. FrankenPHP worker mode is only for HTTP requests, but as FrankenPHP provides a ZTS version of PHP, this paves the way for new usages even in CLI. @alli83, @jwage and I are working in something specifically for Symfony Messenger. Stay tuned!

**Reply by nathan-de-pachtere** [reply]
Hey !

Any updates on the "something specifically for Symfony Messenger" you're working on? 😊

**Reply by alli83** [reply]
Hey!

still under review. @dunglas @nicolas-grekas

**Reply by IndraGunawan** [reply]
hey @alli83 , thanks for your work.

can you share the PR? i would love to see that

**Reply by alli83** [reply]
here is the PR: https://github.com/symfony/symfony/pull/53964

**Reply by withinboredom** [reply]
the original, deleted, comment replied to:

> That PR is about consuming the messages concurrently, is there a way to consume the message using frankenphp on standard way without concurrency? so i can take advantage without waiting the PR to be released
>
> i know i can run the command using supervisord but if there is a frankenphp way it would be great

I'm actually did something like this for $dayJob. It's not open source ... yet (working on it!). But currently it is running in production. You can wire up frankenphp to SQS or RabbitMq and it runs your php concurrently upon receiving a message by wrapping the event in a http request. Hooking it up to other sources is relatively simple, but these are the ones we use at work. It isn't integrated with Symfony or anything like that, but for now, having a worker script boot up the kernel and dispatch is good enough. There are probably existing libraries out there, but it took us about an hour for the initial version and it has slowly grown.

I also have a PoC way to dispatch work between workers. So, you can create two worker pools, one for async work and another for requests. The web workers can then dispatch tasks to the async pool. It's still a super alpha/idea. But you could use this to dispatch emails, do other things, etc. Eventually, I want the API to have an "overflow" where it can dispatch to a global queue (looking at nats because it can be embedded in go) where if the local queue is empty, it can pick up items off the global queue (which may have many, many frontend servers running that can pick up the work). 

This could allow for "packing" quite a bit of compute "on the edge" and then dispatching overflow back to less busy regions via the global queue, at the expense of some latency for the jobs.

There is still a long way to go with this idea though, but I've been working on it off-and-on for a few weeks, between other things.

**Reply by amenophis** [reply]
Hi @withinboredom 👋 
Any news about your implementation ?

**Reply by withinboredom** [reply]
It turns out to be too specific for our use-case to be worth open-sourcing.

**Comment by ruudk** [comment]
I was wondering the same when I was learning about [Temporal for PHP](https://github.com/temporalio/sdk-php). For the worker, it needs [RoadRunner](http://roadrunner.dev) to provide an event loop. With my limited knowledge about the matter I was thinking this is very similar to FrankenPHP's worker mode. Yet that one only triggers on HTTP events. But it would be cool to use things like Temporal workers with FrankenPHP. That would remove the need for RoadRunner and would reduce the number of dependencies. Just have FrankenPHP for your web traffic and asynchronous workloads.
